keel-core/src/main/kotlin/com/netflix/spinnaker/keel/events/PersistentEvent.kt

```diff
@@ -14,13 +14,15 @@ import java.time.Instant
 )
 @JsonSubTypes(
   JsonSubTypes.Type(value = ApplicationEvent::class, name = "application"),
-  JsonSubTypes.Type(value = ResourceEvent::class, name = "resource")
+  JsonSubTypes.Type(value = ResourceEvent::class, name = "resource"),
+  JsonSubTypes.Type(value = PinningEvent::class, name = "pinning")
```

Comment:
New scope for pinning events.

Reply:
what does scope do today in keel in terms of affecting behavior? is it purely informational or does it get used for something?

I ask because ever since the idea of a 'scope' was added I've sort of imagined it as a nice convenient way to serve the filtering/discovery experience we have in mind for the history UI. We think folks will want to jump into an activity feed that is 'scoped' (we've used this term in the design conversations as well) to a particular part of their delivery manifest's structure or domain. For example: show me everything that has happened to this resource (what we show today), show me everything from the `main` environment including the resources inside, or show me everything that's happened to my entire application including the environments/resources inside. While that would theoretically end up surfacing a ton of events at the higher level scopes, we expect to also add event type filtering to let you (or a UI deep link) filter down to only the event types that matter to you in that moment.

If we think the `scope` attribute is appropriate for powering that feature in the future, then it may be worth thinking about changing this event's scope to `environment` since pinning is an action done _to_ an environment _with_ a particular artifact version. If there's other behavioral/architectural needs tied to scope today, then it might be more useful to come up with a separate way of representing the UI/API's idea of 'scoping' later.

Reply:
Good question, @erikmunson. It is not purely informational in the backend. It is used to differentiate different "classes" of events while retaining the ability of using a single `event` table which has the entire history of events in keel that you could sort by various criteria.

Given that, I think it would support _some_ of the use cases you mentioned well, but it currently does not carry the semantics of events that affect an entity and all its "children" (or related/affected entities) as in some of the other examples you give. As I mentioned in our call earlier today, the "meat" of events is actually in the `json` column in the table, so I suspect that, between the `scope`, `timestamp` and `json` columns we should be able to construct queries to cater to all use cases for the UI, and given that MySQL has optimizations for querying into JSON columns, performance should be OK. But I'm sure we'll find out more when we get our hands dirty with it. 🙂

Reply:
Upon discussion with @emjburns I'm removing this from this PR and saving it for a separate one. In any case, the notion of event scopes is already present/usable for resource and application events.